src/py21cmfast/wrapper/outputs.py

```diff
@@ -14,111 +14,525 @@
+            # Compare at the level required by this object only
+            return getattr(self.inputs, f"_{self._compat_hash.name}_hash") == getattr(
+                other, f"_{self._compat_hash.name}_hash"
+            )
+
+        min_req = min(self._compat_hash.value, other._compat_hash.value)
+        min_req = _HashType(min_req)
+
+        return getattr(self.inputs, f"_{min_req.name}_hash") == getattr(
+            other.inputs, f"_{min_req.name}_hash"
+        )
+
+    @property
+    def arrays(self) -> dict[str, Array]:
+        """A dictionary of Array objects whose memory is shared between this object and the C backend."""
+        me = attrs.asdict(self, recurse=False)
+        return {k: x for k, x in me.items() if isinstance(x, Array)}
+
+    @cached_property
+    def struct(self) -> StructWrapper:
+        """The python-wrapped struct associated with this input object."""
+        return StructWrapper(self._name)
+
+    @cached_property
+    def cstruct(self) -> StructWrapper:
+        """The object pointing to the memory accessed by C-code for this struct."""
+        return self.struct.cstruct
+
+    def _init_arrays(self):
+        for k, array in self.arrays.items():
+            # Don't initialize C-based pointers or already-inited stuff, or stuff
+            # that's computed on disk (if it's on disk, accessing the array should
+            # just give the computed version, which is what we would want, not a
+            # zero-inited array).
+            if array.state.c_memory or array.state.initialized or array.state.on_disk:
+                continue
+
+            setattr(self, k, array.initialize())
 
-    _filter_params = _OutputStruct._filter_params + [
-        "T_USE_VELOCITIES",  # bt
-        "MAX_DVDR",  # bt
-    ]
+    @property
+    def random_seed(self) -> int:
+        """The random seed for this particular instance."""
+        return self.inputs.random_seed
+
+    def sync(self):
+        """Sync the current state of the object with the underlying C-struct.
+
+        This will link any memory initialized by numpy in this object with the underlying
+        C-struct, and also update this object with any values computed from within C.
+        """
+        # Initialize all uninitialized arrays.
+        self._init_arrays()
+        for name, array in self.arrays.items():
+            # We do *not* set COMPUTED_ON_DISK items to the C-struct here, because we have no
+            # way of knowing (in this function) what is required to load in, and we don't want
+            # to unnecessarily load things in. We leave it to the user to ensure that all
+            # required arrays are loaded into memory before calling this function.
+            if array.state.initialized:
+                self.struct.expose_to_c(array, name)
+
+        for k in self.struct.primitive_fields:
+            if getattr(self, k) is not None:
+                setattr(self.struct.cstruct, k, getattr(self, k))
+            else:
```

Comment:
Remove else to catch where C and python are both not None and different from each other

Reply:
DOne.